### PR TITLE
Allow PRSS to generate randomness shared with only one helper

### PR DESCRIPTION
**Pull request by akoshelev**
Builds on top of #1185

This closes https://github.com/private-attribution/ipa/issues/1029 by adding a method to PRSS implementation that allows callers to specify which randomness they want to obtain: shared with the helper on the right or on the left. This saves 50% of computation in cases where only one side of PRSS is required. https://github.com/private-attribution/ipa/issues/1029 lists the cases where it is useful.

Happy to bikeshed on naming, it gets harder and harder to find good names

### CODE REVIEW

File: ipa-core/src/protocol/prss/crypto.rs [hunk 1]
```diff
@@ -193,6 +204,27 @@ pub trait SharedRandomness {
...
+    fn generate_one_side<T: FromRandom, I: Into<PrssIndex>>(
+        &self,
+        index: I,
+        direction: Direction,
+    ) -> T {
+        let index = index.into();
+        T::from_random(
+            self.generate_chunks_one_side(index, direction)
+                .next()
+                .unwrap_or_else(|| panic!("Can generate randomness for index {index:?}")),
```
Comment (benjaminsavage) [inline ipa-core/src/protocol/prss/crypto.rs:224]: Can or can't?

**Review by benjaminsavage — APPROVED** [review]

**Comment by akoshelev** [comment]
Here is the link to the benchmark that forced me to implement custom iterator. 


Using `Zip` to combine two iterators. Benchmark against the stable `main`. 15% regression

pdf

Using `ChunksIter` with inline calls. Benchmark against the stable `main`. No perf change 
pdf

File: ipa-core/src/protocol/prss/mod.rs [hunk 2]
```diff
@@ -103,12 +62,25 @@ impl From<PrssIndex128> for u128 {
...
 }
 
+impl TryFrom<u128> for PrssIndex128 {
+    type Error = TryFromIntError;
+
+    fn try_from(value: u128) -> Result<Self, Self::Error> {
+        let value = u64::try_from(value)?;
+
+        Ok(Self {
+            index: u32::try_from(value >> 32).unwrap().into(),
```
Comment (martinthomson) [inline ipa-core/src/protocol/prss/mod.rs:72]:
```suggestion
            index: PrssIndex::from(u32::try_from(value >> 32).unwrap()),
```

**Review by martinthomson — APPROVED** [review]
I had a bunch of comments on the previous one, which you should address there.  Here, the changes look fine.

**Pull request merged**